torch/csrc/autograd/python_variable_indexing.cpp

```diff
@@ -309,111 +267,172 @@ PyObject* THPVariable_getitem(PyObject* self, PyObject* index) {
+
+  // handle simple types: integers, slices
+  if (THPUtils_checkLong(index)) {
+    if (is_tracing && THPVariable_Check(index)) {
+      recordSelectTrace(THPVariable_Unpack(index));
+    }
+    return THPVariable_Wrap(at::indexing::applySelect(self_, 0, THPUtils_unpackLong(index), 0, self_device, self_sizes));
   } else if (PySlice_Check(index)) {
-    return wrap(applySlice(self_, 0, index, true));
+    Py_ssize_t start, stop, step;
+    if (!THPUtils_unpackSlice(index, &start, &stop, &step)) {
+      throw python_error();
+    }
+    if (is_tracing) {
+      Tensor start_tensor, stop_tensor, step_tensor;
+      extractTensorsFromSlice(index, &start_tensor, &stop_tensor, &step_tensor);
+      recordSliceTrace(start_tensor, stop_tensor, step_tensor);
+    }
+    return THPVariable_Wrap(at::indexing::applySlice(
+      self_,
+      0,
+      start,
+      stop,
+      step,
+      /*ensure_view=*/true,
+      /*is_tracing=*/is_tracing,
+      self_device,
+      self_sizes));
   }
 
   // wrap index in a tuple if it's not already one
   THPObjectPtr holder = wrapTuple(index);
 
   variable_list variableIndices;
-  Variable sliced = applySlicing(self_, holder.get(), variableIndices);
+  Variable sliced = applySlicing(self_, holder.get(), variableIndices, is_tracing, self_device, self_sizes);
   if (variableIndices.empty()) {
     if (sliced.is_same(self_)) {
       // ensure we return a shallow copy for things like x[...]
       sliced = at::alias(sliced);
     }
-    return wrap(sliced);
+    return THPVariable_Wrap(sliced);
   }
 
   // indexing by tensors ("advanced" indexing)
-  return wrap(dispatch_index(sliced, variableIndices));
+  return THPVariable_Wrap(([&]() {
+    pybind11::gil_scoped_release no_gil;
+    return at::indexing::dispatch_index(sliced, std::move(variableIndices));
+  })());
+
   Py_RETURN_NONE;
   END_HANDLE_TH_ERRORS
 }
 
-// To match numpy semantics:
-// As a special case for backwards compatibility,
-// strip away unit dimensions from the left of 'src'
-static IntArrayRef slicePrefix1sSize(IntArrayRef sizes) {
-  size_t first_non1_src = sizes.size();
-  for (size_t i = 0; i < sizes.size(); ++i) {
-    if (sizes[i] != 1) {
-      first_non1_src = i;
-      break;
-    }
-  }
-
-  return sizes.slice(first_non1_src);
-}
-
-static void copy_to(Variable dst, const Variable& src) {
-  Tensor b_src;
-  IntArrayRef sliced_src_sizes = slicePrefix1sSize(src.sizes());
-  std::tie(b_src) = expand_inplace(dst, src.view(sliced_src_sizes), "setitem");
-  dst.copy_(b_src);
-}
-
 int THPVariable_setitem(PyObject* self, PyObject* index, PyObject* py_value) {
   HANDLE_TH_ERRORS
   if (py_value == nullptr) {
     throw TypeError("Tensor does not support deleting items");
   }
-
   auto& self_ = reinterpret_cast<THPVariable*>(self)->cdata;
   OptionalDeviceGuard device_guard(device_of(self_));
+  const at::Device& self_device = self_.device();
+  const at::IntArrayRef& self_sizes = self_.sizes();
```

Comment:
You're saved here from use after free by const reference lifetime extension but this is a really dumb way to write this code. Just say `IntArrayRef self_sizes`, the `IntArrayRef` is already a reference type.